### ISSUE TITLE
Cache uniform locations?

### ISSUE THREAD

**Issue by bwasty** [JoeyDeVries/LearnOpenGL#66]
I've been going through the `Shaders` and other tutorials (and [porting](https://github.com/bwasty/learn-opengl-rs) them to Rust) and was slightly suprised that `glGetUniformLocation` is always called directly when setting a uniform (like [here](https://github.com/JoeyDeVries/LearnOpenGL/blob/master/includes/learnopengl/shader_s.h#L81-L94)).
I profiled this a few ago (on an OpenGL 2.1 Desktop app) and found it well worth caching for uniforms that are updated within a render loop.
Might be worth mentioning in the text, or even adding to the `Shader` class?

**Comment by JoeyDeVries** [comment]
Yes, it's indeed more efficient and the standard approach of setting uniforms. However, I tend to write the code as straightforward as possible as to not overwhelm readers with stuff that's not necessary at the time (I usually leave optimization techniques/tricks out). Even though hashing them is quite easy and doesn't generate a lot of code, I'm still more inclined to leave them as is. I eventually want to have a seperate section filled with chapters purely about performance (including tricks to save on GL state changes).
